File: packages/engine/lib/src/planner/load_budget.dart
import 'package:meta/meta.dart';

import '../models/enums.dart';
import 'load_budget_config.dart';

/// The rolling load balance — see `docs/specs/belastungsbudget.md`.
@immutable
class LoadBudget {
  const LoadBudget({required this.quote, required this.recoveryNeed});

  final double quote;
  final RecoveryNeed recoveryNeed;
}

/// Evaluates the rolling load budget from seven days of already-resolved
/// daily loads. Pure function — how a daily load is derived from `Slot`,
/// `Activity.arousal` and `Outcome` is the caller's job, not this
/// function's (see the spec's „Nicht dazu gehört").
LoadBudget evaluateLoadBudget({
  required List<int> loadOverLastSevenDays,
  required LifeStage lifeStage,
  required Set<Restriction> restrictions,
  required LoadBudgetConfig config,
}) {
  assert(
    loadOverLastSevenDays.length == 7,
    'evaluateLoadBudget expects exactly 7 days, got ${loadOverLastSevenDays.length}',
  );

  var capacity = config.capacityPerDay[lifeStage]!;
  for (final restriction in restrictions) {
    final cap = config.restrictionCap[restriction];
    if (cap != null && cap < capacity) capacity = cap;
  }

  final totalLoad = loadOverLastSevenDays.fold(0, (sum, day) => sum + day);
  final quote = totalLoad / 7 / capacity;

  final recoveryNeed = quote >= config.recoveryNeedHighFrom
      ? RecoveryNeed.high
      : quote >= config.recoveryNeedMediumFrom
          ? RecoveryNeed.medium
          : RecoveryNeed.none;

  return LoadBudget(quote: quote, recoveryNeed: recoveryNeed);
}

File: infra/supabase/functions/_shared/planner/steps/load_budget.ts
import type { LifeStage, RecoveryNeed, Restriction } from '../models/enums.ts';
import type { LoadBudgetConfig } from './load_budget_config.ts';

/** The rolling load balance — see `docs/specs/belastungsbudget.md`. */
export interface LoadBudget {
  readonly quote: number;
  readonly recoveryNeed: RecoveryNeed;
}

/**
 * Evaluates the rolling load budget from seven days of already-resolved
 * daily loads. Pure function — how a daily load is derived from `Slot`,
 * `Activity.arousal` and `Outcome` is the caller's job, not this
 * function's (see the spec's „Nicht dazu gehört").
 */
export function evaluateLoadBudget(args: {
  loadOverLastSevenDays: readonly number[];
  lifeStage: LifeStage;
  restrictions: ReadonlySet<Restriction>;
  config: LoadBudgetConfig;
}): LoadBudget {
  const { loadOverLastSevenDays, lifeStage, restrictions, config } = args;
  if (loadOverLastSevenDays.length !== 7) {
    throw new Error(
      `evaluateLoadBudget expects exactly 7 days, got ${loadOverLastSevenDays.length}`,
    );
  }

  const baseCapacity = config.capacityPerDay.get(lifeStage);
  if (baseCapacity === undefined) {
    throw new Error(`load budget config has no capacity entry for life stage ${lifeStage}`);
  }
  let capacity = baseCapacity;
  for (const restriction of restrictions) {
    const cap = config.restrictionCap.get(restriction);
    if (cap !== undefined && cap < capacity) capacity = cap;
  }

  const totalLoad = loadOverLastSevenDays.reduce((sum, day) => sum + day, 0);
  const quote = totalLoad / 7 / capacity;

  const recoveryNeed: RecoveryNeed = quote >= config.recoveryNeedHighFrom
    ? 'high'
    : quote >= config.recoveryNeedMediumFrom
    ? 'medium'
    : 'none';

  return { quote, recoveryNeed };
}
